### ISSUE TITLE
[HOLD for payment 2024-06-20] [HOLD for payment 2024-06-18] [$250] [TS Migration] Standartize between Boolean or !!

### ISSUE THREAD

**Issue by mountiny** [Expensify/App#39126]
Follow up issue for TS migration project. Coming from [this spreadsheet](https://docs.google.com/spreadsheets/d/1GVp81ixTZ3TIspkTsgfhSA72KZGHDJlwRa4J8TtCQU4/edit#gid=1579985166)

> Standartize between Boolean or !!

Lets pick the one which can be enforced with eslint

Any more details for this one cc @blazejkustra @fabioh8010
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01111a2979f5f2ac8c</li>
        <li>Upwork Job ID: 1780333266166345728</li>
        <li>Last Price Increase: 2024-04-16</li>
<li>Automatic offers: </li>
<ul>
<li>ishpaul777 | Contributor | 0</li>
<li>ShridharGoel | Contributor | 0</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @CortneyOfstad</details>

**Comment by blazejkustra** [comment]
@ShridharGoel Not yet, let's wait until migration is fully finished

**Comment by fabioh8010** [comment]
This is on hold Melvin..

**Comment by blazejkustra** [comment]
**Details:**

We want to standardize with just one approach, !! and Boolean() are almost identical and we want to ensure just one is used in the codebase to avoid cases like this 😅
```ts
offline = Boolean(network.shouldForceOffline) || !!network.isOffline;
```

This includes writing a proposal on Slack, adjusting the code in the repository and perhaps adding an eslint rule that ensures only one approach is used. I myself believe !! is more readable more concise and faster in runtime.

**This issue requires a minimal understanding of how Expensify process works, before coding the contributor have to post a Problem/Solution on open-source channel**

**Comment by blazejkustra** [comment]
The issue is still on hold, please wait until it's open for external contributors @ShridharGoel

**Comment by mountiny** [comment]
In the proposal for this issue please write the slack problem/ solution statement to present in favour of the option you want to standardize on as Blazej mentioned:

> This includes writing a proposal on Slack, adjusting the code in the repository and perhaps adding an eslint rule that ensures only one approach is used. I myself believe !! is more readable more concise and faster in runtime.

**Comment by roryabraham** [comment]
> I myself believe !! is ... faster in runtime

@blazejkustra what makes you say this?

**Comment by blazejkustra** [comment]
Read some articles and benchmarks on the topic, sorry for not linking before:
- [Benchmark](https://www.measurethat.net/Benchmarks/Show/11127/0/boolean-vs)
- [Boolean() or !! (double bang, double negation)?](https://gist.github.com/arthurvi/66cb1e2bcfc92f99f465e0db04264367#performance)
 

Also some interesting posts here:
- [Boolean doesn't narrow down type the same way as !! does](https://stackoverflow.com/questions/66788887/react-typescript-why-does-boolean-behave-differently-than-double-not-op)
```ts
/* eslint-disable react/function-component-definition */
import React from 'react';

type Props = {
    myNumber: number | null;
};

const Test: React.FC<Props> = ({myNumber}) => (
    <>
        {/* Example 1 */}
        {!!myNumber && <span>{myNumber.toString()}</span>}

        {/* Example 2 */}
        {Boolean(myNumber) && <span>{myNumber.toString()}</span>} // error here
    </>
);

export default Test;
```
- [Some positive opinions on reddit in favor of `Boolean`](https://www.reddit.com/r/typescript/comments/oty3jl/opinions_on/)

**Comment by saifelance** [comment]
> Triggered auto assignment to @CortneyOfstad (`NewFeature`), see https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=BugZero%20process%20steps%20for%20feature%20requests for more details. Please add this Feature request to a GH project, as outlined in [the SO](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=Add%20the%20issue%20to%20a%20GH%20project).



> Triggered auto assignment to Contributor-plus team member for initial proposal review - @ntdiary (`External`)

@MelvinBot 

// ESLint rule for enforcing Boolean usage
module.exports = {
    create: function(context) {
        return {
            UnaryExpression: function(node) {
                if (node.operator === "!" && node.argument.type === "UnaryExpression" && node.argument.operator === "!") {
                    context.report({
                        node: node,
                        message: "Use Boolean instead of !!",
                        fix: function(fixer) {
                            const sourceCode = context.getSourceCode();
                            const start = node.range[0];
                            const end = node.argument.range[1];
                            const argumentText = sourceCode.getText(node.argument);
                            return fixer.replaceTextRange([start, end], `Boolean(${argumentText})`);
                        }
                    });
                }
            }
        };
    }
};


 if the decision is to standardize on !!, the ESLint rule should be adjusted accordingly:
 
 
 // ESLint rule for enforcing !! usage
module.exports = {
    create: function(context) {
        return {
            CallExpression: function(node) {
                if (node.callee.type === "Identifier" && node.callee.name === "Boolean" && node.arguments.length === 1) {
                    context.report({
                        node: node,
                        message: "Use !! instead of Boolean",
                        fix: function(fixer) {
                            const sourceCode = context.getSourceCode();
                            const argumentText = sourceCode.getText(node.arguments[0]);
                            return fixer.replaceText(node, `!!${argumentText}`);
                        }
                    });
                }
            }
        };
    }
};

**Comment by saifelance** [comment]
https://www.upwork.com/freelancers/saifurrehman

**Comment by saifelance** [comment]
***@***.***
https://www.upwork.com/freelancers/saifurrehman

**Comment by saifelance** [comment]
> Started the discussion on this: https://expensify.enterprise.slack.com/archives/C01GTK53T8Q/p1713329747825759

Unable to access, I think need to allow permission?

**Comment by saifelance** [comment]
proposal sent

**Comment by CortneyOfstad** [comment]
@ntdiary when you have a chance, can you review the proposal above and provide any feedback? Thank you!

**Comment by CortneyOfstad** [comment]
No worries! Feel free to post and and we can get that adjusted!

**Comment by CortneyOfstad** [comment]
I agree @ishpaul777 — and thanks for taking this on!

**Comment by saifelance** [comment]
> I dont have strong opinions, but i like consizeness with `!!` over `Boolean()` if we put it up for vote i'd 👍 for`!!`

Yes. Another solution is to standardize the usage of !!. In this scenario, the ESLint check should ensure that !! is consistently used instead of Boolean.

Depending on the decision, appropriate changes should be made to the ESLint configuration and rules.

**Comment by saifelance** [comment]
@MelvinBot @CortneyOfstad I sent above both solutions code updates, it was ignore or rejected work?

**Comment by saifelance** [comment]
@MelvinBot I sent my proposal on Upwork job, did you check?

**Comment by saifelance** [comment]
Kindly send Slack invite chanel : ***@***.***

**Comment by roryabraham** [comment]
I'm not sure we had broad enough consensus yet, posted a proposal [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1713547892853379)

**Comment by mountiny** [comment]
Seems like we are now good to move ahead with !!

**Comment by CortneyOfstad** [comment]
@ishpaul777 can you provide a follow-up on the proposal that fits the clarification on `!!`? If we can get an assignment ready  to roll, that would be great 👍

**Comment by roryabraham** [comment]
@ShridharGoel when do you estimate you'll have a PR ready for review?

**Comment by roryabraham** [comment]
thanks for the update!

**Comment by CortneyOfstad** [comment]
Any update @ishpaul777 on the PR review? Thanks!

**Comment by CortneyOfstad** [comment]
Thanks @ishpaul777!

**Comment by fabioh8010** [comment]
@aldo-expensify kindly bump 🙂

**Comment by CortneyOfstad** [comment]
Not overdue — just waiting for @aldo-expensify to review PR [here](https://github.com/Expensify/eslint-config-expensify/pull/96) 👍

**Comment by aldo-expensify** [comment]
There is a problem with the commits signing, left comment in the PR

**Comment by CortneyOfstad** [comment]
@ShridharGoel it looks like there is some outstanding comments on the PR. Can you please have those addressed by EOD today? Thanks!

**Comment by CortneyOfstad** [comment]
Freeze will be done on Wednesday evening, so will re-address then!

**Comment by CortneyOfstad** [comment]
@ishpaul777 Unfortunately, the overdue label cannot be removed until @ShridharGoel comments as they have the star indicating action is waiting on them. 

@ShridharGoel can you please provide the PR by EOD?

**Comment by CortneyOfstad** [comment]
@ShridharGoel bump on the update please

**Comment by CortneyOfstad** [comment]
PR is being actively reviewed so no additional update at this stage!

**Comment by CortneyOfstad** [comment]
@ishpaul777 can you confirm if this needs a regression test please before EOD to help avoid a delay in payment? Thank you!

**Comment by CortneyOfstad** [comment]
Bump @ishpaul777 ^^^

**Comment by CortneyOfstad** [comment]
@ShridharGoel I had to reinvite you to the job post because for some reason you were not listed in Upwork — let me know once you accept and I can get that paid ASAP!

**Comment by CortneyOfstad** [comment]
@ShridharGoel let me know once you accept and I can get this paid ASAP. Thanks!

**Comment by CortneyOfstad** [comment]
## Payment Summary

@ishpaul777 — paid $250 via Upwork
@ShridharGoel — paid $250 via Upwork
